### ISSUE TITLE
Low Level 1-D and 2-D Array Operations

### ISSUE THREAD

**Issue by damianmoz** [chapel-lang/chapel#16200]
During a project where Chapel is being used, the manipulation of 1D-arrays (vectors), or 1D-slices of 2D-arrays (a matrix) saw various issues raise their heads. For example, given
```chapel
	var y : [someRange] real;
	var x : [someRange] real;
        var z : [someRange, someRange] real;
```
Statements like
```chapel
	[r in someRange] y += x[r] * z[r, ..];
```
performs sub-optimally, while iterating over the domain of the matrix was a more performant solution, albiet with less clarity. That is not a complaint, just a comment. Such operations might get vectorized in future, but not for now. The overhead of doing this operation in parallel in some scenarios is often too high and in other scenarios not feasible.

Various other things were noted. **Unrolling loops** to some degree results in better performance. The limited results from this project showed that an unrolling factor of four (4) is optimal.  Anything more had too much overhead. It was found that using a param index variable to force the unrolling was a waste of time. **Slicing an array** before passing it to a routine was more effective than passing the slice directly as an argument and using it as such. And vica versa. Are there guidelines for this? **Using complex numbers** to simplify the algebra of a planar rotation for the optimizer was slower than giving the optimizer the complex multiplication longhand to attack.

In an effort to better hide the peculiarities, do we need to have a Chapel variant of what Fortran called BLAS. Over 40 years since that library was published, overloaded function names can be exploited to provide a cleaner and more concise version of this type of thing. There are Level-10 operations which have 1D array input and scalar (0d) output, There are Level-11 which have 1D arrays as input and 1D arrays as input/output. There are Level-12 which have 1D arrays as input and 2D arrays as input/output, Level 21 with up to 2D-arrays as input and 1D arrays as input/output, and finally Level-22 which have 2D arrays as input and 2D arrays as input/output.

This would be native Chapel, not some interface to BLAS or LAPACK, and not some heavily tiled thing that is more suited to huge arrays. There may be work involved in reworking, enhancing, or augmenting, the algorithms and the BLAS interface to better work with Chapel's column-major array storage. When the native Chapel code achieves something close to what the optimized versions of the C-Blas interfaces do, Chapel will have certainly arrived.  Not that I am proposing something which achieves performance of that class now. The use of such an interface in the code which is the expression of the algorithm will loose some clarity. But that can be offset somewhat by a judiciously named interface, albeit not entirely.

There are issues which will influence this work. These include the resolution of #16007 and #14364, one of which is talking about coercion and one about casting, the provision of a reliable, optimal and predictable use of an FMA instruction to ensure numerical accuracy in some operations, and the availability of a true **vector** type as in #15812, not the glorified list in #15668.

Comments? Feedback?

Worth pursuing or a waste of time? If the former, what are the design considerations, architecture, etc?

**Comment by bradcray** [comment]
Hi Damian — Thanks for taking the time to put these notes together.  I think a native-Chapel BLAS (or BLAS-like) library would be very attractive from the perspective of demonstrating maturity of the language and serving as a forcing function for digging into node-level optimizations that we might otherwise put off longer.  I think it'd also be interesting to compare its productivity/elegance/clarity/maintainability to the standard BLAS library (where, as you note, there'd likely be a "most elegant" Chapel version that may not be competitive today, but ought to be someday; and a "most performant" Chapel version that would perform more competitively, but not be as pretty to look at.  Ideally these two converge into the best of both worlds over time.

Now the real question: Are you proposing authoring such a library?  :)

Just to note, I feel like there is some overlap between this effort and the LinearAlgebra library, though I think the approach taken there is a mix of wrapping existing libraries and creating native Chapel versions (so it's not as "pure" of an approach, arguably).  I think (?) that the scope of what it's trying to express is also broader than what you're proposing here.  But I wouldn't want to get too far with any native Chapel BLAS-like effort without coordinating with those working on the LA library to make sure no toes are stepped on or efforts were wasted.  Tagging @ben-albrecht for that reason.

**Comment by damianmoz** [comment]
It is called the Basic Limear Algebra Module or **BLAM**. I am having finger/mouse trouble attaching a draft report for this so will send it to you to attach if you think it is useful as it stands. That report is so, so, crude at this stage.

The answer to the question is that I have done some of it already, but only a small part of it, and very roughly, and it needs to be redone. I had to do it for the work on our current Singular Value Decomposition (SVD) project (which had lots of other goals as well).  To do BLAM properly needs an implementation of that **vector** module (or ideally a **vector** type in Chapel that does the same).  There is also access to an FMA instruction which Chapel is missing as far as I can tell. Happy to be told to RTFM on that. I will probably need that partial reduce working as well. Also, lots more stuff to learn. So the answer is yes, but not today, oir even next month. We will need guidance and also read my next paragraph where I reply to your last paragraph.

I have seen the Liner Algebra stuff. I was after a _pure approach as that integrates with our R+D direction so it does not really fit. Wrapping stuff is not part of that direction. Not sure of my timeline as my main focus is elsewhere. In the absence of a decent **vector** module, we may be forced to wrap new C code because I can find nothing which fits except for some of the excellent lower level C++ template stuff in EIGEN (but again, that is C++).

And when I have that BLAM report at revision 0.09, have wrapped up the almost finished SVD project and addressed questions raised by it, and a whole lot of other stuff,I will pester @ben-albrecht. So maybe September. If there is a detailed specification document and plan associated with the Linear Algebra Module, that might make useful bed-time reading.

**Comment by damianmoz** [comment]
[blam-002.pdf](https://github.com/chapel-lang/chapel/files/5039830/blam-002.pdf)

Some draft documentation is attached. Draft with a capital D.

**Comment by ben-albrecht** [comment]
> I feel like there is some overlap between this effort and the LinearAlgebra library, though I think the approach taken there is a mix of wrapping existing libraries and creating native Chapel versions (so it's not as "pure" of an approach, arguably)

> I have seen the Liner Algebra stuff. I was after a _pure approach as that integrates with our R+D direction so it does not really fit. Wrapping stuff is not part of that direction

Just to clarify on this - the [LinearAlgebra module](https://chapel-lang.org/docs/master/modules/packages/LinearAlgebra.html) utilizes both wrapping existing libraries and native implementations, where the user can switch which implementation they wish to use based on if the dependencies are available on their system.

We currently have a GSoC student extending the Linear Algebra module native implementation, who happens to be working on SVD right now (draft PR opened yesterday https://github.com/chapel-lang/chapel/pull/16205). Maybe there is an opportunity for some collaboration there. @damianmoz - Is your work-in-progress SVD implementation available somewhere?

> To do BLAM properly needs an implementation of that vector module (or ideally a vector type in Chapel that does the same). 

The LinearAlgebra module opted to use native Chapel arrays to represent vectors and matrices because Chapel arrays come with a rich interface that can be extended with open methods (e.g. LinearAlgebra adds `_array.T` for transpose).  That said, we have and continue to consider modifying the interface and introducing _vector_ and _matrix_ types. This idea has not received much attention lately though.

> But I wouldn't want to get too far with any native Chapel BLAS-like effort without coordinating with those working on the LA library to make sure no toes are stepped on or efforts were wasted. 

From what I understand, BLAM aims to be a lower-level library (similar to BLAS/LAPACK) than what the LinearAlgebra library is designed to be. I think progress in both projects could benefit from each other. Depending on its success, BLAM could potentially be another backend for the LinearAlgebra library, or we could consider porting some BLAM implementations to the LinearAlgebra library and vice versa.

**Comment by damianmoz** [comment]
> Just to clarify on this - the LinearAlgebra module utilizes both wrapping existing libraries and native implementations,
> where user can switch which implementation they wish to use based on if the dependencies are available on their system.

I saw LinearAlgebra as soon as it appeared and have reviewed it on each new chapel release. Wrapping is not our research  direction. I was interested in its native Chapel implementation. But, it appears that LinearAlgebra is more targetting the ease of expression of what a MATLAB or SCILAB provides (but correct me if I am wrong). It did not seem to fit our needs. Also, I noticed that it did not exploit many of the techniques learned from the development of libraries like LINPACK, LAPACK, and more recently, PLASMA, MAGMA, FLAME and EIGEN. This was a bit disappointing. Or am I looking at an old version? Are there design documents or discussion papers for LinearAlgebra?

Thanks for that information about the GSoC student working on SVD. I had to go fishing for that because it does  not show up on Github for mortals. I had a quick look and got dismayed. Why is he looking at a 'pythag' function?  What reference text is he using to gain an understanding of how to program SVD?

I am happy to collaborate. Our work-in-progress is not publically available but I will make it so in the next two weeks. Sorry, too many meetings next week. My day job has way too much management. The SVD code is not yet as well documented as it will be and it needs to change anyway  What reference texts on SVD do you have?

Please remember that the BLAM report has come from a project whose primary aim was to provide programming guidelines for Chapel parallel algorithms for use in partial differential equation codes. It was deliberately constrained by a need to produce a highly readable implementation of the candidate algorithm, the SVD.  Part of that same project was to ascertain how much (or little) of a BLAS-like library was needed for production applications. It was hoped that BLAM would be a lot smaller than the draft report seems to indicate. That is not a complaint, just a comment.  In any case, BLAM is still at the design stage.

> The LinearAlgebra module opted to use native Chapel arrays to represent vectors and matrices because Chapel arrays
> come with a rich interface that can be extended with open methods (e.g. LinearAlgebra adds _array.T for transpose).
> That said, we have and continue to consider modifying the interface and introducing vector and matrix types. This idea
> has not received much attention lately though.

Chapel's arrays are more than adequate for the task to which you refer and I would not recommend introducing vector and matrix types. Keep it simple. If you think you need to, stop and revisit (and maybe enhance) Chapel arrays.

When I used the word 'vector' earlier, I was mathematically incorrect but consistent #15812. Therein, 'vector' is used in the context of 'vector processing'. That discussion concerns a module which exposes a data type which is a short sequence of numeric data which can be manipulated by so-called vector instructions. That module avoids (or hides) architecture-dependent 'vector' intrinsics. In that context, a 'vector' is nothing like some special 1D-array and should never appear in high level code, let alone a user interface. The name of  the #15812 data type needs collaborative  jam-sessions but the result is unlikely to be 'vector'. The noun 'vector'got hijacked too many times during the last half of the 20th century. I do not want to perpetuate it into the 21st century although some others keep trying to!

BLAM can be viewed as BLAS with a focus on column-major operations. It is very low level. No LAPACK interfaces. I think there are two tiny pieces of LAPACK's eigenvalue computation which need to go into BLAM. One is about upodating a norm, well a sum of squares because BLAM assumes a norm is a Just-in-Time computation.The other operation has slipped my mind.

I certainly agree that progress in both projects could benefit from each other.

> Depending on its success, BLAM could potentially be another backend for the LinearAlgebra library,
> or we could consider porting some BLAM implementations to the LinearAlgebra library and vice versa.

Sure. Note that if it was felt there was a need to port something from BLAM, that likely means that BLAM's design is inadequate in some area and needs to be corrected. It is bound to have holes in it.

**Comment by ben-albrecht** [comment]
>  it appears that LinearAlgebra is more targetting the ease of expression of what a MATLAB or SCILAB provides (but correct me if I am wrong)

That is absolutely correct. It's aimed to be a high-level interface to linear algebra operations.

> Also, I noticed that it did not exploit many of the techniques learned from the development of libraries like LINPACK, LAPACK, and more recently, PLASMA, MAGMA, FLAME and EIGEN. This was a bit disappointing.

The development process of the LinearAlgebra library has been loosely iterating on the following cycle:

1. Wrap existing libraries where we can to benefit from the years of research and development that went into those libraries with relatively little effort.
2. Write native implementations of operations from (1), without worrying too much about performance yet (performant implementations are available via the wrapped implementations).
3. Come back and spend effort optimizing implementations from (2)

With the vast feature space of all Linear Algebra operations x (dense, sparse) x (local, distributed, gpu), most of the effort has gone into steps (1) and (2) so far, striving for feature completeness first, native implementation performance second.

Because we only wrap LAPACK and BLAS today, sparse and distributed implementations begin with step (2), though we expect to support wrapped sparse, distributed, and gpu implementations eventually.

> Are there design documents or discussion papers for LinearAlgebra?

Not really beyond the information presented in various GitHub issues and the module documentation. For example, here's an issue describing our goal for local dense feature completeness: https://github.com/chapel-lang/chapel/issues/15621

> Thanks for that information about the GSoC student working on SVD. I had to go fishing for that because it does not show up on Github for mortals. I had a quick look and got dismayed. Why is he looking at a 'pythag' function? What reference text is he using to gain an understanding of how to program SVD?

For the initial local implementation, Rahul is following the Numerical Recipes implementation (chapter 2.6). We're also looking at [Handbook on Parallel Computing and Statistics](https://www.irisa.fr/sage/bernard/publis/SVD-Chapter06.pdf) for more performant implementations in the future. Which references are you following for your implementation?

>  What reference texts on SVD do you have?

See above.

> I am happy to collaborate. Our work-in-progress is not publically available but I will make it so in the next two weeks.

Sounds great. Let me know when it's available. Are you pursuing a local or distributed implementation? Taking a look at the Draft, it appears BLAM is targeting local (non-distributed) operations initially. Is that correct?

> When I used the word 'vector' earlier, I was mathematically incorrect but consistent #15812. Therein, 'vector' is used in the context of 'vector processing'.

Ah, I was interpreting your request as the combination of a mathematical vector type that is optimized with vector-processing operations. That makes sense. I agree the term "vector" is a bit overloaded. I support the idea of giving users more control in this regard, but I don't have a strong opinion on whether it should be exposed through a module interface or language construct. (maybe that discussion should take place over on #15812 though).

> Sure. Note that if it was felt there was a need to port something from BLAM, that likely means that BLAM's design is inadequate in some area and needs to be corrected. It is bound to have holes in it.

Understood, perhaps the LinearAlgebra feature set can help guide BLAM's design as well then.

**Comment by damianmoz** [comment]
Thanks for the information about #15621.  You are a man of very few words. I will need to connect a lot of dots and go and refer to SciPy. I will look at the LinearAlgebra feature set in more detail. I missed it when it first appeared on GitHub although I do not diligently sit camped on Github.

As far as documentation about the NR variant of SVD, I found the original version of the Golub-Reinsch paper useful. Note that they do some naughty things in it like reuse data for the wrong purpose and rely on sentinels. It would not pass a static analysis. Many of the performance inprovements that are raised in the Handbook document are implemented clearly in the GSL (GNU Scientific Library).

We are only looking at a local implementation.  As I probably mentioned, our aims of doing an SVD went beyond SVD as a piece of Linear Algebra. Our focus was parallel programming guidelines for local algorithms so going distributed did not fit with the aims. It I went for a distributed version, I would look beyond the pure Golub-Reinsch implementation.

I believe that there should be a predefined type in the language to allow users to get at the 'vector' instructions. But doing a module means less work for the Chapel team for now and a chance to iron out the issues. Once that is bedded down, then implement a Chapel type to do the work. But, yes, it belongs in #15812 which I started when I found somebody trying to use the word 'vector' to mean a specialized list which is certainly not what it is.

On the GSoc student, ignore 'pythag' as seen in the NR variant of SVD. Either expose 'hypot' from LIBC and use that or, more look at what is going on and create a Givens' planar rotation (like that below as per the Linpack algorithm). Ed Anderson's recent (2017) paper has a superior implementation but it is more complicated and needs extra constants. I note that I have what I consider to be an improvement which is not yet published and will not be for quite a few months yet. I do believe that both hypot and an algorithm like *RotG belong in LibC but there are so many conflicting ways to implement this that even if it does get included, it will be a mess. Also a generalization really needs the ability to return a tuple, something C lacks.  I do not use the IEEE754 standard hypot myself because it yields the wrong answer for my purposes when one of the arguments is a NaN and the other is an Infinity. But that is a whole other argument that has gone on for decades both privately and in public.
```chapel
// compute Givens Plane rotation by BLAS-1 (LINPACK)
// Needs to be replaced with Anderson's Improvements

inline proc vmRotG(p : ?Real, q : Real) : (Real, Real, Real)
{
        const zero = 0:Real;
        const one = 1:Real;
        const ap = abs(p);
        const aq = abs(q);
        var r, c, s : Real;

        if ap + aq == zero then
        {
                (r, c, s) = (zero, one, zero);
        }
        else
        {
                const tau = ap + aq;
                const (tp, tq) = (p / tau, q / tau);
                const t = sqrt(tp * tp + tq * tq);

                r = if (if ap > aq then p else q) < zero then -t else +t;
                (r, c, s) = ( r * tau, tp / r, tq / r);
        }
        return (r, c, s);
}
```

**Comment by ben-albrecht** [comment]
> Thanks for the information about #15621. You are a man of very few words. I will need to connect a lot of dots and go and refer to SciPy. I will look at the LinearAlgebra feature set in more detail. I missed it when it first appeared on GitHub although I do not diligently sit camped on Github.

I think it'd be great to put out a public document describing the design goals of the Linear Algebra module - it's just not something we typically have done for modules (especially non-standard library modules), with exception of a few modules originally proposed as [CHIPs](https://github.com/chapel-lang/chapel/tree/master/doc/rst/developer/chips). I am happy to answer any questions that you have. Knowing what design aspects are unclear to an outside contributor will help us know what details should be necessary in the formal design document if we eventually put effort into that.

Thanks for the SVD resources. I'll share with our student and project co-mentors. As mentioned before, our first implementation is focusing on correctness / simplicity, where we can pursue more performant (and complex) algorithms as a follow-up effort. We chose the NR implementation as this starting point.

**Comment by damianmoz** [comment]
The NR implementation should  be OK to start. Incorporating the Given's rotation code I previously emailed addresses the major inaccuracy source that the NR code has.

What interface with your SVD routine e expose?  Will you preallocate the arrays or do you propose to allocate the factorization within the routine? 

I will try and get my code into a form it can be shared. Its contents and what we learned doing it will hopefully be the subject of a session at CHIUW 2020.

**Comment by ben-albrecht** [comment]
@damianmoz - Thanks. The Given's rotation changes seem like a clear next step.

> What interface with your SVD routine e expose? Will you preallocate the arrays or do you propose to allocate the factorization within the routine?

Current interface will match that of the existing [LAPACK-wrapped svd](https://chapel-lang.org/docs/master/modules/packages/LinearAlgebra.html#LinearAlgebra.svd), which sacrifices performance (allocates arrays within the function) for simplicity of the interface. We'd eventually like to extend this interface (along with many LinearAlgebra interfaces) to support passing in any temp or return arrays that need to be allocated to let users manage that themselves. This will be important for the distributed interface too, since the return array will let the user specify the resulting distribution.


> I will try and get my code into a form it can be shared. Its contents and what we learned doing it will hopefully be the subject of a session at CHIUW 2020.

Cool, looking forward to it! I think you mean CHIUW 2021 (typically in May)?
